[Xing Zheng] iP

xzzz3 opened the pull request:
# XzzzBot
### A project based on the original DukeBot

> "It takes as much energy to wish as it does to plan." -- Eleanor Roosevelt ([source](https://lifehacker.com/it-takes-as-much-energy-to-wish-as-it-does-to-plan-5825679))

XzzzBot helps you to remember what you can't be bothered to! 
It's:
- text-based
- easy to learn and use
- *SUPER* efficient
- ~~specifically programmed to sound like a zoomer to make you feel young and entertained~~

AND....... IT'S **FREE**!!!

All you need to do to start using XzzzBot is:
1. download it from [here](https://github.com/xzzz3/ip),
2. double-click the downloaded file,
3. add your tasks,
4. relax and let XzzzBot manage your tasks for you 😆 ! 

Some of the features:
- [x] Managing tasks
- [ ] Managing different types of tasks like events, deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, feel free to use it to practice Java and improve the bot! Here's the `main` method in the `Duke` class:
```
public static void main(String[] args) {
    new Duke().run();
}
```

Commit:
Add Gradle support

xzzz3 committed:
Initialized for Level-1, added basic customization and responses.

xzzz3 committed:
Included an ArrayList for level-2, and added the Add and Display functionality.

xzzz3 committed:
Cleaning up code with helper functions and switch statements.

xzzz3 committed:
Added Task class to assist with the marking, as well as the mark function.

xzzz3 committed:
Added inherited classes: todo, deadlines and events.

xzzz3 committed:
Modified the respond function to make it work with different task types, as well as added more lively responses.

xzzz3 committed:
Done TextUiTesting, updated the behaviour and edited the expected file to the updated and correct behaviour.

xzzz3 committed:
Added several error classes to handle different errors, caused by the three task types.

xzzz3 committed:
Added method for deleting tasks from the list.

xzzz3 committed:
Minor improvements by removing unnecessary toString methods.

xzzz3 committed:
Switch Case statements are not in the convention form.

It should adhere to the indentation convention.

Let's:
* Edit the indentation to fit the conventional standard.

xzzz3 committed:
Add test file for future testing

Currently lacking a test file

We need a test file for future testing to be done

Let's:
* add a test file

xzzz3 committed:
Add toText function for tasks to facilitate file saving

The toString function is too wordy for file-saving in the harddrive.
The amount of words also make it difficult for reading later on.

Let's:
* Add a toText method to generate a concise summary of the task

xzzz3 committed:
Ignore data file that saves the list

The data file is always changing in the hard drive as the list changes.

It should not be tracked on git as the changes are not significant.

Let's:
* Ignore the data file and add it to gitignore.

xzzz3 committed:
Add file access functions

File saving to hard drive is needed for the list.

Let's:
* Add read and write functions to the chatbot so that it can read
and modify the text file that stores the list.
* Invoke these functions at the start and end of each session to
update the text file.

xzzz3 committed:
Deadline and Event classes: Add date parsing functionality

The deadline and event classes take in only string that are date
descriptions. They cannot translate numerical date values in string.

Sometimes this functionality will be needed as it is more machine
friendly and may be preferred by people.

Let's:
* Change the constructors to check for parse-able date strings
* Parse the date strings if applicable
* Use the parsed date string as the stored date instead

xzzz3 committed:
Main Duke class: Trim input

The user input does not have the space trimmed for date values. For
instance, evaluating the date of "/by 5pm" gives " 5pm" with a space
at the start, which leads to unexpected behaviors.

Let's:
* Trim the string whenever applicable to remove leading spaces

xzzz3 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Event.java

xzzz3 committed:
Ui Class: Add new Ui Class to consolidate the ui functions

The output functions are done using system.output within the Main
file, which was not fully adhering to OOP standards.

Let's:
* Create a new Ui class to consolidate all the Ui functions.
* Edit the original Main class to substitute previous output functions
to Ui class methods.
* Remove exceptions that are now unnecessary with the addition of a
Ui class.

xzzz3 committed:
Storage class: Add storage class and supporting methods

The chatbot requires a storage class to deal with loading tasks from
the file and saving tasks in the file.

A separate class is preferred for OOP standards.

Let's:
* Create the storage class.
* Add LoadingExceptions class for storage class methods to throw.
* Edit the Ui class to work with storage class.

xzzz3 committed:
Parser class: Add new Parser class

Require a Parser class to deal with making sense of the user command.

Let's:
* Add a Parser class to parse user commands, when the command needs
to be split into multiple parts.

Additional Remarks: Did not quite get what a parser class is useful
for, as most parts of the code are rather blended together and cannot
be shifted to Parser. May require future changes to perfect Parser.

xzzz3 committed:
TaskList Class: Add new TaskList Class to store the task list.

A TaskList class is required to contain the tasklist itself, so as
to separate the tasklist from the chatbot implementation.

Let's:
* Create the new TaskList class.
* Add basic functionalities.

xzzz3 committed:
Main Duke class: Update Functionalities with the new OOP additions.

The class is outdated with the new OOP additions which can be used
to substitute some of the chatbot's code.

Let's:
* Move relevant parts to the other classes.
* Debug the Duke class with the new addition.
* Clean up.

xzzz3 committed:
All classes: Package into duke.test package

Classes are scattered. They should be consolidated into a package
for ease of use and for good practice.

Let's:
* Package all of them except main into duke.test package
* Import the package in the main Duke class
* Move them to the right location in the hard drive

Note that in the future there may be other packages based on the use.

xzzz3 committed:
Parser Class: Move some codes in main Duke class to Parser class

Some code that are more suitable for Parser class to handle are in
the main Duke class.

This can potentially cause dependability, as the main class is handling
some of the Parser class's codes.

Let's:
* Move the relevant parts to Parser class, mainly the exception
checking and trimming of results
* Ensure that the result passed back to the main Duke class is
processed and ready for use, which corresponds to the purpose of a
Parser class.
* Add a new exception IncompleteArgumentException for Parser exceptions.

xzzz3 committed:
Junit test files: Add test files

The classes require unit testing with test files.

This is to ensure ease of testing in the future and also a consolidated
testing platform.

Let's:
* Add test files for Parser and ToDo
* Add relevant testing methods for different scenarios
* Check that all of them are functioning as expected

xzzz3 committed:
Add META-INF/MANIFEST.MF to gitignore

xzzz3 committed:
All Classes: Add JavaDoc comments

The classes only includes the methods. Without JavaDoc comments,
it will be difficult for external users to understand the purpose of
these methods.

Let's:
* Add JavaDoc comments to all the methods in all classes.

xzzz3 committed:
Storage class: Remove unnecessary full method names

Some of the methods are called using their full package names in
the storage class, like duke.task.Deadline().

This is wordy and can lead to messy code.

Let's:
* Remove these instances of full names.

xzzz3 committed:
Duke, TaskList and Ui class: Add functionality for Find function

The chatbot requires a find functionality that outputs matching file
names.

This can help the end user find the interested files more easily.

Let's:
* In Duke class, add response trigger to "find" command.
* In TaskList class, add method to find the relevant tasks.
* In Ui class, add method to output the relevant tasks found.

xzzz3 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/Storage.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/task/Ui.java

Resolved all the conflicts

xzzz3 committed:
Task class: Change Boolean attribute done to isDone

The done attribute as a Boolean do not follow the coding standards,
which is that they should sound like the English language when
evaluated to true or false.

Let's:
* Change the Done attribute to isDone in Task.
* Change all instances where Done is used to isDone

lzan98 commented on src/main/java/duke/task/Duke.java:
```diff
@@ -0,0 +1,157 @@
+        respond();
+
+        storage.saveFile(taskList.getToDoList());
+    }
```
I like how the run() method is neat and tidy by separating the logic into another method.

lzan98 commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,100 @@
+                case "E":
+                    toDoList.add(new Event(taskLine[2], taskLine[3]));
+                    break;
+                }
```
Perhaps a default case is missing in this switch block?

xzzz3 replied:
Thank you for catching that! I will update it shortly.

lzan98 commented on src/main/java/duke/task/Duke.java:
```diff
@@ -0,0 +1,157 @@
+                ui.showFindResult(relevantTasks);
+                break;
+            default:
+                ui.doNotUnderstand();
```
I like how you made use of switch statements instead of 'if-else', which made the code much neater than mine.

lzan98 commented on src/main/java/duke/task/Ui.java:
```diff
@@ -0,0 +1,139 @@
+     * @param toDoList List of tasks.
+     * @param idx Index of the task that has been marked.
+     */
+    public void markAsDone(ArrayList<Task> toDoList, int idx) {
+        System.out.println("yay!!! this task is now marked as done ٩(˘◡˘)۶");
+        System.out.println(toDoList.get(idx - 1).toString());
+    }
+
+    /**
+     * Prints a confirmation that a <code>Task</code> at the index in the list has been marked as undone.
+     * @param toDoList List of tasks.
+     * @param idx Index of the task that has been unmarked.
+     */
+    public void unmarkAsDone(ArrayList<Task> toDoList, int idx) {
+        System.out.println("this task is now marked as not done yet... do it soon! ᕙ(`▿´)ᕗ");
+        System.out.println(toDoList.get(idx - 1).toString());
+    }
```
Perhaps I missed it out while reading your code, but was the possibility that the user enters an invalid task number for mark/unmark considered? (e.g. 3 tasks in the list but user enters mark 10).

xzzz3 replied:
Hmm... You bring up a good point! I forgot to consider cases where the idx exceeds the size or when the idx is negative. Thanks!

lzan98 commented on src/main/java/duke/task/Ui.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+
+    /**
+     * Prints the list for display.
+     * @param toDoList List to be displayed.
+     */
+    public void displayList(ArrayList<Task> toDoList) {
+        int number = 1;
+        System.out.println("here are your tasks ☜(ˆ▿ˆc)");
+        for (Task item : toDoList) {
+            System.out.println(number + ". " + item.toString());
+            number++;
+        }
+    }
```
Another good usage of OOP here. Well done~

lzan98 reviewed (commented):
Very neat and tidy code with good usage of OOP, as well as JavaDoc comments. No coding standard violations were found.

xzzz3 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

xzzz3 committed:
Gradle files: Add Gradle to the project

Gradle can help to automate the build process as well as other
processes such as unit testing for the project. Thus, adding gradle
will be an important improvement for the project.

Let's:
* Merge the gradle support branch
* Initialize the gradle file

xzzz3 committed:
Checkstyle files: Add checkstyle plugin

The project code has no checkstyle and only relies on the IDE's
prompts for code formatting. A checkstyle plugin can help to improve
the consistency and correctness of the code.

Let's:
* Add the checkstyle plug in from IntelliJ's marketplace
* Use the plugin with project checkstyle files

xzzz3 committed:
All Java classes: Edit code with checkstyle

Some parts of the code have been tagged problematic with the new
checkstyle plugin.

Let's:
* Use the checkstyle hints to improve the coding quality
* Minor fixes to the syntax for some classes

xzzz3 committed:
GUI setup classes: Add GUI basics to XzzzBot

The chat bot is more complete with a GUI instead of a text-based UI.

Let's:
* Follow the tutorial and add basic GUI to the bot.
* Continue to work on more features to make the GUI have the correct
functionality.

xzzz3 committed:
GUI and Ui related files: Update text Ui to become GUI

The UI of XzzzBot is a text-based UI, and can be improved to a GUI
with the newly added FXML. This can improve the user experience.

Let's:
* Update the text-based printing methods in the Ui class to accommodate
GUI functionalities
* Remove certain uncommon symbols that cannot be represented on GUI
* Update the main Duke class to work with the FXML files

joeietao commented on src/main/java/duke/task/Duke.java:
```diff
@@ -0,0 +1,165 @@
+        }
+
+        switch (command) {
+        case "bye":
```
I like the switch statement, it makes the code easy to understand and concise!

joeietao commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * Represents a task. A <code>Task</code> object is represented by its name and a boolean indicating whether it is done.
+ */
+public abstract class Task {
```
I like the naming of your methods and variables here, they are very intuitive!

joeietao commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+        System.out.println(by);
+        try {
+            LocalDate d1 = LocalDate.parse(by);
+            this.by = d1.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
```
maybe I didn't understand your code properly but I'm not very sure why the date format is MMM d yyyy HAHA

joeietao reviewed (commented):
Overall I think your code was very clean and easy to understand, there's nothing much to catch HAHA good job!!

xzzz3 committed:
Duke, Storage, TaskList, ToDo, Ui Classes: Add Assert functions

There are certain states of the program that are assumed to hold true.
By using assertions, these states can be checked, in case the functions
work unexpectedly, or when the future changes affect the current
assumptions.

Let's:
* Add assert statements to check state of the program
* Clean up some code with the use of assert statements

xzzz3 committed:
Duke, MainWindow, TaskList, Ui Classes: Review code quality

Some of the codes did not follow the code quality, such as verb-sounding
methods.

By reviewing and changing them, the code can be more readable for readers
and more understandable for collaborators.

Let's:
* Change the method names to fit the coding conventions
* Change the switch-case statement to be clearer visually

xzzz3 committed:
Merge pull request #1 from xzzz3/branch-A-Assertions

Add Branch-A-assertions

xzzz3 committed:
Merge branch 'branch-A-CodeQuality'

xzzz3 committed:
Task sub-classes: Update DateTime object to store date instead of String

The tasks with specified dates are currently storing them as strings
instead of date objects. While this is easier to implement, it makes
them harder to compare.

For the Increment on detecting date anomalies, such comparison is
required and thus the date formats need to be updated accordingly.

Let's:
* Change the stored date's representations to LocalDateTime
* Update functionalities

xzzz3 committed:
Task Classes (mainly Event class): Add detection for date clashes

Sometimes users may store events that happen at the same time. These
should be immediately reflected to them as they cannot do two events
at the same time.

Let's:
* Add a date detection function in the TaskList class
* Modify the other classes to respond to situations where the event
clashes

xzzz3 committed:
README.md: Update User Guide (UG)

For Xzzzbot to be used by others, a UG is necessary to provide a
quick summary of all the commands and features.

Let's:
* Update the README file to include a summarized UG.
* Fix some Ui bugs found along the way.

xzzz3 committed:
Set theme jekyll-theme-cayman

xzzz3 committed:
JavaFX related classes (Main, MainWindow):

Some of the expected functions like closing after saying "bye" is
missing. Those can help to make the user experience better.

Let's:
* Add in close and save functionalities.

xzzz3 committed:
Merge branch 'master' of https://github.com/xzzz3/ip
